csrc/executor_kernel_arg.cpp

```diff
@@ -14,6 +14,292 @@
 
 namespace nvfuser {
 
+namespace {
+
+// Forward traverse from rFactor domain to allocation domain, compute frontier
+// sizes and strides, validate that splits are divisible and merges are
+// contiguous, and update active_ids_ correspondingly.
+class ForwardTraverseFromRFactorToAlloc {
+  ExpressionEvaluator& ee_;
+  std::unordered_map<IterDomain*, std::pair<int64_t, int64_t>>& active_ids_;
+
+  void handle(Split* split) {
+    auto in = split->in();
+    auto inner = split->inner();
+    auto outer = split->outer();
+    auto in_it = active_ids_.find(in);
+    // TORCH_INTERNAL_ASSERT(in_it != active_ids_.end())
+    if (in_it == active_ids_.end()) {
+      // TODO: see [Allocation domain on both side of rFactor]
+      return;
+    }
+    auto [in_size, in_stride] = in_it->second;
+    auto factor = ee_.evaluate(split->factor())->as<int64_t>();
+    TORCH_INTERNAL_ASSERT(
+        in_size % factor == 0,
+        "non-divisible split is not allowed in allocation domain");
+    TORCH_INTERNAL_ASSERT(active_ids_.erase(in) == 1);
+    TORCH_INTERNAL_ASSERT(
+        active_ids_
+            .emplace(inner, std::pair<int64_t, int64_t>{factor, in_stride})
+            .second);
+    TORCH_INTERNAL_ASSERT(active_ids_
+                              .emplace(
+                                  outer,
+                                  std::pair<int64_t, int64_t>{
+                                      in_size / factor, in_stride * factor})
+                              .second);
+  }
+
+  void handle(Merge* merge) {
+    auto inner = merge->inner();
+    auto outer = merge->outer();
+    auto out = merge->out();
+    auto inner_it = active_ids_.find(inner);
+    auto outer_it = active_ids_.find(outer);
+    // TORCH_INTERNAL_ASSERT(inner_it != active_ids_.end())
+    // TORCH_INTERNAL_ASSERT(outer_it != active_ids_.end())
+    if (inner_it == active_ids_.end() || outer_it == active_ids_.end()) {
+      // TODO: see [Allocation domain on both side of rFactor]
+      return;
+    }
+    auto [inner_size, inner_stride] = inner_it->second;
+    auto [outer_size, outer_stride] = outer_it->second;
+    TORCH_INTERNAL_ASSERT(
+        inner_stride * inner_size == outer_stride,
+        "Merging of discontiguous dimensions is not allowed in allocation domain");
```

Comment:
Why isn't it allowed?

Reply:
If you merge two IDs into one, these two IDs must be contiguous. This is just like, if you have a tensor whose shape is `(3, 4)`, you can view it as `(12,)` only if it is contiguous. Otherwise 
```
for i in range(12):
  index = i * stride[0]
```
will fail.

Reply:
Maybe a better way to explain this is, if you have a tensor of shape `(3, 4)`, you can allocate a flat size 12 tensor to store this `(3, 4)` tensor if and only if the `3` is contiguous w.r.t. the `4`.

Reply:
Makes sense.